### PR TITLE
Support vendor verify

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
The scripts is copied from @liggitt [kubernetes pr ](https://github.com/kubernetes/kubernetes/pull/74877) and make a small modification

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
the `jq` is not installed, @howardjohn should i install it here `https://github.com/istio/test-infra/blob/a9ed9324824392f6554bf681872d449f81dcb3c0/docker/istio/istio/install.sh`?

**Comment by hzxuzhonghu** [comment]
opened https://github.com/istio/test-infra/pull/1364/

**Comment by geeknoid** [comment]
I don't really want code to verify that vendor is correct, I want to get rid of the vendor directory completely in the coming weeks. It is completely redundant and is largely deprecated given go modules. We're already not using it in our other repos.

File: vendor/modules.txt [hunk 1]
```diff
@@ -169,474 +169,473 @@ github.com/go-redis/redis/internal/pool
...
 github.com/open-policy-agent/opa/storage
 github.com/open-policy-agent/opa/storage/inmem
 github.com/open-policy-agent/opa/topdown
 github.com/open-policy-agent/opa/topdown/builtins
-# github.com/openshift/api v0.0.0-20190322043348-8741ff068a47
+github.com/open-policy-agent/opa/types
+github.com/open-policy-agent/opa/util
+# github.com/openshift/api v0.0.0-20190322043348-8741ff068a47 => github.com/openshift/api v0.0.0-20190322043348-8741ff068a47
 github.com/openshift/api/apps/v1
-# github.com/opentracing/opentracing-go v1.0.2
```
Comment (howardjohn) [inline vendor/modules.txt]: This doesn't seem right, why is everything changing?

Comment (hzxuzhonghu) [reply vendor/modules.txt]: It does not change, 

```

# ensure_require_replace_directives_for_all_dependencies:
# - ensures all existing 'require' directives have an associated 'replace' directive pinning a version
# - adds explicit 'require' directives for all transitive dependencies
# - adds explicit 'replace' directives for all require directives (existing 'replace' directives take precedence)
function ensure_require_replace_directives_for_all_dependencies() {
```

**Comment by hzxuzhonghu** [comment]
@geeknoid You can not assume every one have very well network environment to get all the vendors.

**Comment by hzxuzhonghu** [comment]
@howardjohn Now the test-infra image has been updated, can I think the vendor check can pass now.

**Comment by hzxuzhonghu** [comment]
@howardjohn @geeknoid what's your concern? The pr is conflicting frequently, I have rebased many times.

**Comment by howardjohn** [comment]
I don't know why we need all this complexity. Why can't we just do `go mod tidy && go mod vendor&& git status`? That is what we did with dep.

I think Martin already gave his opinion
> I don't really want code to verify that vendor is correct, I want to get rid of the vendor directory completely in the coming weeks. It is completely redundant and is largely deprecated given go modules. We're already not using it in our other repos.

there is already some work to go in this direction too.

for what its worth, personally, I don't care much whether or not we vendor.

**Comment by hzxuzhonghu** [comment]
> Why can't we just do go mod tidy && go mod vendor&& git status? 

That is basically verfy-vendor does.

In addition, this pr provides scripts to upgrade pkg and update go mod, thus users do not need care much about go mod commands

**Comment by hzxuzhonghu** [comment]
As can see, by running update-vendor.sh, it prunes some useless files under `vendor/github.com/lestrrat-go/jwx/jwt/`

**Comment by howardjohn** [comment]
this just seems like a huge hack around go modules and it's not clear to me what benefit we are getting for all the complexity.

**Comment by hzxuzhonghu** [comment]
suppose one has to upgrade a vendor, how should he do 
1. set Env xxxxx  
1. exec go get xxxxx
1. go mod vendor
1. maybe think of exec `go mod tidy `

How we can make sure all does this in order, how to check it. Though one can even edit go.mod manually. How can you prevent it?

**Comment by howardjohn** [comment]
I would much rather run `go get istio.io/api@latest`, which is universal across go projects then have to find some script and run `./updateVendors.sh istio.io/api@latest`. We can check they did it correctly with the lint that just does `go mod tidy && go mod vendor&& git status` which if you try will catch the same `github.com/lestrrat-go/jwx` issue

File: bin/pin-dependency.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 The Istio Authors.
```
Comment (ericvn) [inline bin/pin-dependency.sh]: Not sure if we can simply pull code with a Kubernetes Authors copyright and replace it with an Istio authors one? This is pervasive across the files.

File: bin/pin-dependency.sh [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+ISTIO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+
+# Usage:
+#   hack/pin-dependency.sh $MODULE $SHA-OR-TAG
```
Comment (ericvn) [inline bin/pin-dependency.sh]: Multiple places where _hack_ has been changed to _bin_ in the directory structure, but not in the comments.

**Comment by hzxuzhonghu** [comment]
Given that this may have a license issue, i agree we add a simple script for this.

**Comment by hzxuzhonghu** [comment]
I have no bandwidth on this, let me close it.

**Pull request closed without merging**